packages/jsii-diff/lib/classes-ifaces.ts

```diff
@@ -0,0 +1,203 @@
+}
+
+/**
+ * Find the indicated parameter with the given index
+ *
+ * May return the last parameter if it's variadic
+ */
+function findParam(parameters: reflect.Parameter[], i: number): reflect.Parameter | undefined {
+  if (i < parameters.length) { return parameters[i]; }
+  const lastParam = parameters.length > 0 ? parameters[parameters.length - 1] : undefined;
+  if (lastParam && lastParam.variadic) { return lastParam; }
+  return undefined;
+}
+
+function compareProperty(origClass: reflect.Type, original: reflect.Property, updated: reflect.Property, context: ComparisonContext) {
+  if (original.static !== updated.static) {
+    // tslint:disable-next-line:max-line-length
+    context.mismatches.report(origClass, `property ${original.name}, used to be ${original.static ? 'static' : 'not static'}, is now ${updated.static ? 'static' : 'not static'}`);
+  }
+
+  const ana = isStrengtheningPostCondition(original.type, updated.type);
+  if (!ana.success) {
+    context.mismatches.report(origClass, `property ${original.name}, type ${describeTypeMatchingFailure(original.type, updated.type, ana)}`);
+  }
+
+  if (updated.immutable && !original.immutable) {
+    context.mismatches.report(origClass, `property ${original.name}, used to be mutable, is now immutable`);
+  }
+}
+
+// tslint:disable-next-line:max-line-length
+function* memberPairs<T extends reflect.TypeMember>(origClass: reflect.ClassType | reflect.InterfaceType, xs: T[], updatedClass: reflect.ClassType | reflect.InterfaceType, context: ComparisonContext): IterableIterator<[T, T]> {
+  for (const origMember of xs.filter(shouldInspect(context))) {
+    LOG.trace(`${origClass.fqn}#${origMember.name}`);
+
+    const updatedMember = updatedClass.members.find(m => m.name === origMember.name);
+    if (!updatedMember) {
+      context.mismatches.report(origClass, `member ${origMember.name} has been removed`);
+      continue;
+    }
+
+    if (origMember.kind !== updatedMember.kind) {
+      context.mismatches.report(origClass, `member ${origMember.name} changed from ${origMember.kind} to ${updatedMember.kind}`);
+    }
+
+    if (!origMember.protected && updatedMember.protected) {
+      context.mismatches.report(origClass, `member ${origMember.name} changed from 'public' to 'protected'`);
+    }
+
+    yield [origMember, updatedMember as T]; // Trust me I know what I'm doing
+  }
+}
+
+/**
+ * Whether we are strengthening the postcondition (output type of a method or property)
+ *
+ * Strengthening postconditions is allowed!
+ */
+function isStrengtheningPostCondition(original: reflect.TypeReference, updated: reflect.TypeReference): Analysis {
+  if (original.void) { return { success: true }; }  // If we didn't use to return anything, returning something now is fine
+  if (updated.void) { return { success: false, reasons: [`now returning 'void'`] }; } // If we used to return something, we can't stop doing that
+  return isSuperType(original, updated, updated.system);
```

Comment:
Why not move the void checks to isSuperType and get rid of there two methods? I feel that the weakening/strengthening semantics add unneeded additional cognitive load

Reply:
They are not the same, the order is reversed.